File: Hard/154_Find_Minimum_in_Rotated_Sorted_Array_II/Solution.java


public class Solution {
    public int findMin(int[] nums) {
        int left = 0;
        int right = nums.length - 1;

        while (left < right) {
            int mid = left + (right - left) / 2;

            if (nums[mid] > nums[right]) {
                left = mid + 1;
            } else if ((nums[mid] < nums[right])) {
                right = mid;
            } else {
                right--;
            }
        }

        return nums[left];
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        int[] nums1 = {1,3,5};
        System.out.println(solution.findMin(nums1));

        // test cases 2
        int[] nums2 = {2,2,2,0,1};
        System.out.println(solution.findMin(nums2));
    }
}

File: Hard/154_Find_Minimum_in_Rotated_Sorted_Array_II/Solution.py
from typing import List, Optional


class Solution:
    def findMin(self, nums: List[int]) -> int:
        left = 0
        right = len(nums) - 1

        while (left < right):
            mid = left + (right - left) // 2

            if (nums[mid] > nums[right]):
                left = mid + 1
            elif (nums[mid] < nums[right]):
                right = mid
            else:
                right -= 1
        
        return nums[left]


if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    nums1 = [1,3,5]
    print(solution.findMin(nums1))

    # test cases 2
    nums2 = [2,2,2,0,1]
    print(solution.findMin(nums2))
    